hw_diag/utilities/hardware.py

```diff
@@ -373,6 +380,97 @@ def is_nebra_device(diagnostics: dict) -> bool:
     return "nebra" in friendly_name.lower()
 
 
+def has_external_antenna_support() -> bool:
+    device_type = os.getenv('BALENA_DEVICE_TYPE')
+    return device_type in EXT_ANT_DEVICE_TYPES
+
+
+def is_external_antenna_enabled() -> bool:
+    bc = balena_cloud.BalenaCloud.new_from_env()
+    variables = bc.get_device_config_variables()
+    for var_dict in variables:
+        if var_dict['name'] not in DTPARAM_CONFIG_VAR_NAMES:
+            continue
+
+        # dtparam var value is a comma-separated list of values to be used as separate
+        # `dtparam=<val>` lines in `config.txt`
+        return EXT_ANT_DTPARAM in var_dict['value']
+
+    return False
+
+
+def _get_device_dtparams_var(bc: balena_cloud.BalenaCloud) -> tuple[list[str], int]:
+    # BALENA_HOST_CONFIG_dtparam/RESIN_HOST_CONFIG_dtparam var value is a comma-separated list of
+    # values to be used as separate `dtparam=<val>` lines in `config.txt`.
+    variables = bc.get_device_config_variables()
+    for var_dict in variables:
+        if var_dict['name'] not in DTPARAM_CONFIG_VAR_NAMES:
+            continue
+
+        device_dtparams_var_id = var_dict['id']
+
+        value = var_dict['value'].strip()
+        if value:
+            device_dtparams = value.split(',')
+        else:
+            device_dtparams = []
+        break
+    else:
+        device_dtparams = []
+        device_dtparams_var_id = None
+
+    return device_dtparams, device_dtparams_var_id
+
+
+def _get_fleet_dtparams_var(bc: balena_cloud.BalenaCloud) -> list[str]:
+    variables = bc.get_fleet_config_variables()
+    for var_dict in variables:
+        if var_dict['name'] not in DTPARAM_CONFIG_VAR_NAMES:
+            continue
+
+        value = var_dict['value'].strip()
+        if value:
+            return value.split(',')
+        else:
+            return []
+
+    return []
+
+
+def set_external_antenna_enabled(enabled: bool) -> None:
+    bc = balena_cloud.BalenaCloud.new_from_env()
+
+    # The device variable overrides the fleet variable, so we need to deal with both.
+    device_dtparams, device_dtparams_var_id = _get_device_dtparams_var(bc)
+    fleet_dtparams = _get_fleet_dtparams_var(bc)
+
+    if enabled:
```

Comment:
the complexity of the function can be reduced by breaking out parameter preparation.

Reply:
This is the only actual logic inside the function and it's tightly coupled with creating/updating the device config variable. Therefore it doesn't make too much sense to pull parameter preparation out, IMHO.